File: quickstart/iOS-swift/ZUMOAPPNAME/ZUMOAPPNAME/QSTodoListViewController.m
// ----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// ----------------------------------------------------------------------------
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#import <WindowsAzureMobileServices/WindowsAzureMobileServices.h>
#import "QSTodoListViewController.h"
#import "QSTodoService.h"


#pragma mark * Private Interface


@interface QSTodoListViewController ()

// Private properties
@property (strong, nonatomic)   QSTodoService   *todoService;
@property (nonatomic)           BOOL            useRefreshControl;

@end


#pragma mark * Implementation


@implementation QSTodoListViewController

@synthesize todoService;
@synthesize itemText;
@synthesize activityIndicator;


#pragma mark * UIView methods


- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Create the todoService - this creates the Mobile Service client inside the wrapped service
    self.todoService = [QSTodoService defaultService];
    
    // Set the busy method
    UIActivityIndicatorView *indicator = self.activityIndicator;
    self.todoService.busyUpdate = ^(BOOL busy)
    {
        if (busy)
        {
            [indicator startAnimating];
        } else
        {
            [indicator stopAnimating];
        }
    };
    
    // add the refresh control to the table (iOS6+ only)
    [self addRefreshControl];
    
    // load the data
    [self refresh];
}

- (void) refresh
{
    // only activate the refresh control if the feature is available
    if (self.useRefreshControl == YES) {
        [self.refreshControl beginRefreshing];
    }
    [self.todoService refreshDataOnSuccess:^
    {
        if (self.useRefreshControl == YES) {
            [self.refreshControl endRefreshing];
        }
        [self.tableView reloadData];
    }];
}


#pragma mark * UITableView methods


- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
 forRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Find item that was commited for editing (completed)
    NSDictionary *item = [self.todoService.items objectAtIndex:indexPath.row];
    
    // Change the appearance to look greyed out until we remove the item
    UILabel *label = (UILabel *)[[tableView cellForRowAtIndexPath:indexPath] viewWithTag:1];
    label.textColor = [UIColor grayColor];
    
    // Ask the todoService to set the item's complete value to YES, and remove the row if successful
    [self.todoService completeItem:item completion:^(NSUInteger index)
    {  
        // Remove the row from the UITableView
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index inSection:0];
        [self.tableView deleteRowsAtIndexPaths:@[ indexPath ]
                              withRowAnimation:UITableViewRowAnimationTop];
    }];
}

-(UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Find the item that is about to be edited
    NSDictionary *item = [self.todoService.items objectAtIndex:indexPath.row];
    
    // If the item is complete, then this is just pending upload. Editing is not allowed
    if ([[item objectForKey:@"complete"] boolValue])
    {
        return UITableViewCellEditingStyleNone;
    }
    
    // Otherwise, allow the delete button to appear
    return UITableViewCellEditingStyleDelete;
}

-(NSString *)tableView:(UITableView *)tableView titleForDeleteConfirmationButtonForRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Customize the Delete button to say "complete"
    return @"complete";
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    
    // Set the label on the cell and make sure the label color is black (in case this cell
    // has been reused and was previously greyed out
    UILabel *label = (UILabel *)[cell viewWithTag:1];
    label.textColor = [UIColor blackColor];
    NSDictionary *item = [self.todoService.items objectAtIndex:indexPath.row];
    label.text = [item objectForKey:@"text"];
    
    return cell;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Always a single section
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of items in the todoService items array
    return [self.todoService.items count];
}


#pragma mark * UITextFieldDelegate methods


-(BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [textField resignFirstResponder];
    return YES;
}


#pragma mark * UI Actions


- (IBAction)onAdd:(id)sender
{
    if (itemText.text.length  == 0)
    {
        return;
    }
    
    NSDictionary *item = @{ @"text" : itemText.text, @"complete" : @NO };
    UITableView *view = self.tableView;
    [self.todoService addItem:item completion:^(NSUInteger index)
    {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index inSection:0];
        [view insertRowsAtIndexPaths:@[ indexPath ]
                    withRowAnimation:UITableViewRowAnimationTop];
    }];
    
    itemText.text = @"";
}


#pragma mark * iOS Specific Code

// This method will add the UIRefreshControl to the table view if
// it is available, ie, we are running on iOS 6+

- (void)addRefreshControl
{
    Class refreshControlClass = NSClassFromString(@"UIRefreshControl");
    if (refreshControlClass != nil)
    {
        // the refresh control is available, let's add it
        self.refreshControl = [[UIRefreshControl alloc] init];
        [self.refreshControl addTarget:self
                                action:@selector(onRefresh:)
                      forControlEvents:UIControlEventValueChanged];
        self.useRefreshControl = YES;
    }
}

- (void)onRefresh:(id) sender
{
    [self refresh];
}


@end

File: quickstart/iOS-swift/ZUMOAPPNAME/ZUMOAPPNAME/QSTodoListViewController.swift
// ----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// ----------------------------------------------------------------------------
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import UIKit

class QSTodoListViewController : UITableViewController {
    
    @IBOutlet weak var itemText : UITextField
    @IBOutlet weak var activityIndicator : UIActivityIndicatorView

    var todoService: QSTodoService = QSTodoService.defaultService
    var useRefreshControl: Bool = false
    
    // #pragma mark * UIView methods
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Set the busy method
        var indicator = activityIndicator

        todoService.busyUpdate = {(busy:Bool) in
            if (busy) {
                indicator.startAnimating()
            } else {
                indicator.stopAnimating()
            }
        }
        
        // add the refresh control to the table (iOS6+ only)
        addRefreshControl()
        
        // load the data
        refresh()
    }
    
    func refresh() {
        // only activate the refresh control if the feature is available
        if (useRefreshControl == true) {
            refreshControl.beginRefreshing()
        }
        
        todoService.refreshDataOnSuccess({() in
            if (self.useRefreshControl == true) {
                self.refreshControl.endRefreshing()
            }
            self.tableView.reloadData()
        })
    }

    // #pragma mark * UITableView methods
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        // Find item that was commited for editing (completed)
        let item = todoService.items[indexPath.row] as NSDictionary
        
        // Change the appearance to look greyed out until we remove the item
        let label = tableView.cellForRowAtIndexPath(indexPath).viewWithTag(1) as UILabel
        label.textColor = UIColor.grayColor()
        
        // Ask the todoService to set the item's complete value to YES, and remove the row if successful
        todoService.completeItem(item, completion:{index in
            let indexPath = NSIndexPath(forRow:index, inSection:0)
            tableView.deleteRowsAtIndexPaths([indexPath] , withRowAnimation:UITableViewRowAnimation.Top)
        })
    }

    override func tableView(tableView: UITableView, editingStyleForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCellEditingStyle {
        // Find the item that is about to be edited
        let item = todoService.items[indexPath.row] as NSDictionary
        
        // If the item is complete, then this is just pending upload. Editing is not allowed
        if (item["complete"].boolValue == true) {
            return UITableViewCellEditingStyle.None;
        }
        
        // Otherwise, allow the delete button to appear
        return UITableViewCellEditingStyle.Delete;
    }
    
    override func tableView(tableView: UITableView, titleForDeleteConfirmationButtonForRowAtIndexPath indexPath: NSIndexPath) -> String! {
        // Customize the Delete button to say "complete"
        return "complete"
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let CellIdentifier = "Cell"
        var cell = tableView.dequeueReusableCellWithIdentifier(CellIdentifier) as UITableViewCell
        if (cell == nil) {
            cell = UITableViewCell(style: UITableViewCellStyle.Default, reuseIdentifier:CellIdentifier)
        }
        
        // Set the label on the cell and make sure the label color is black (in case this cell
        // has been reused and was previously greyed out
        let label = cell.viewWithTag(1) as UILabel
        label.textColor = UIColor.blackColor()
        let item = todoService.items[indexPath.row] as NSDictionary
        label.text = item["text"] as String
        
        return cell;
    }
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> NSInteger {
        // Always a single section
        return 1;
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection selection: NSInteger) -> NSInteger {
        // Return the number of items in the todoService items array
        return todoService.items.count
    }

    
    // #pragma mark * UITextFieldDelegate methods
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }

    
    // #pragma mark * UI Actions
    
    @IBAction func onAdd(sender: AnyObject) {
        if (itemText.text.bridgeToObjectiveC().length  == 0) {
            return
        }
        
        let item: Dictionary = ["text" : itemText.text as String, "complete" : false ]
        let view = tableView;
        todoService.addItem(item, completion:{index in
            let indexPath = NSIndexPath(forRow:index, inSection:0)
            view.insertRowsAtIndexPaths([indexPath], withRowAnimation:UITableViewRowAnimation.Top)
        })
        itemText.text = "";
    }

    // #pragma mark * iOS Specific Code
    
    // This method will add the UIRefreshControl to the table view if
    // it is available, ie, we are running on iOS 6+
    
    func addRefreshControl() {
        var refreshControlClass:AnyClass = NSClassFromString("UIRefreshControl")
        if (refreshControlClass != nil) {
            // the refresh control is available, let's add it
            refreshControl = UIRefreshControl()
            refreshControl.addTarget(self, action: "onRefresh:", forControlEvents: UIControlEvents.ValueChanged)
            useRefreshControl = true
        }
    }
    
    func onRefresh(sender: AnyObject) {
        refresh()
    }
}